bd-logger/src/global_state.rs

```diff
@@ -0,0 +1,101 @@
+// shared-core - bitdrift's common client/server libraries
+// Copyright Bitdrift, Inc. All rights reserved.
+//
+// Use of this source code is governed by a source available license that can be found in the
+// LICENSE file or at:
+// https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt
+
+#[cfg(test)]
+#[path = "./global_state_test.rs"]
+mod tests;
+
+use bd_device::Store;
+use bd_key_value::{Key, Storable};
+use bd_log_primitives::{LogField, LogFields};
+use std::sync::Arc;
+
+const KEY: Key<State> = Key::new("global_state");
+
+#[derive(Debug, serde::Serialize, serde::Deserialize, Default, PartialEq, Eq)]
+struct State(Vec<(String, String)>);
```

Comment:
This is going to get checked a lot, I would probably switch this to a HashMap for quicker lookup in the common case since we can't guarantee the order of the incoming.

Reply:
I had that originally but changed it back to this since I realized that we haven't deduped logs at this point and I didn't want to have to construct a new HashMap just to make the comparison. Since the metadata providers are all evaluated in order it seems like we can rely on the order being consistent, but maybe this is a good time to switch over to using a map type everywhere?

Reply:
Of course that was a ton of work, probably worth another cursory review